webapp/src/ts/services/enketo.service.ts

```diff
@@ -288,44 +289,63 @@ export class EnketoService {
       });
   }
 
-  private renderFromXmls(doc, wrapper, instanceData) {
+  private renderFromXmls(xmlFormContext: XmlFormContext) {
+    const { wrapper } = xmlFormContext;
     wrapper
       .find('.form-footer')
       .addClass('end')
       .find('.previous-page,.next-page')
       .addClass('disabled');
 
     const formContainer = wrapper.find('.container').first();
-    formContainer.html(doc.html);
+    formContainer.html(xmlFormContext.doc.html.get(0));
+
+    return this.getEnketoOptions(xmlFormContext.doc, xmlFormContext.instanceData)
+      .then((options) => {
+        this.currentForm = new window.EnketoForm(wrapper.find('form').first(), options);
+        const loadErrors = this.currentForm.init();
+        if (loadErrors && loadErrors.length) {
+          return Promise.reject(new Error(JSON.stringify(loadErrors)));
+        }
+      })
+      .then(() => this.getFormTitle(xmlFormContext.titleKey, xmlFormContext.doc))
+      .then((title) => {
+        this.setFormTitle(wrapper, title);
+        wrapper.show();
 
-    return this.getEnketoOptions(doc, instanceData).then((options) => {
-      this.currentForm = new window.EnketoForm(wrapper.find('form').first(), options);
-      const loadErrors = this.currentForm.init();
-      if (loadErrors && loadErrors.length) {
-        return Promise.reject(new Error(JSON.stringify(loadErrors)));
-      }
-      // manually translate the title as enketo-core doesn't have any way to do this
-      // https://github.com/enketo/enketo-core/issues/405
-      const $title = wrapper.find('#form-title');
-      if (doc.title) {
-        // title defined in the doc - overwrite contents
-        $title.text(this.translateFromService.get(doc.title));
-      } else if ($title.text() === 'No Title') {
-        // useless enketo default - remove it
-        $title.remove();
-      } // else the title is hardcoded in the form definition - leave it alone
-      wrapper.show();
-
-      wrapper.find('input').on('keydown', this.handleKeypressOnInputField);
-
-      // handle page turning using browser history
-      window.history.replaceState({ enketo_page_number: 0 }, '');
-      this.overrideNavigationButtons(this.currentForm, wrapper);
-      this.addPopStateHandler(this.currentForm, wrapper);
-      this.forceRecalculate(this.currentForm);
-
-      return this.currentForm;
-    });
+        wrapper.find('input').on('keydown', this.handleKeypressOnInputField);
+
+        // handle page turning using browser history
+        window.history.replaceState({ enketo_page_number: 0 }, '');
+        this.overrideNavigationButtons(this.currentForm, wrapper);
+        this.addPopStateHandler(this.currentForm, wrapper);
+        this.forceRecalculate(this.currentForm);
+
+        return this.currentForm;
+      });
+  }
+
+  private getFormTitle(titleKey, doc) {
+    if (titleKey) {
+      // using translation key
+      return this.translateService.get(titleKey);
+    } else if (doc.title) {
```

Comment:
You don't need the `else` here, since you're already returning in the previous block.